extensions/data-plane/data-plane-aws-s3/src/main/java/org/eclipse/edc/connector/dataplane/aws/s3/S3DataSink.java

```diff
@@ -52,37 +52,41 @@ protected StreamResult<Object> transferParts(List<DataSource.Part> parts) {
                 var partNumber = 1;
                 var completedParts = new ArrayList<CompletedPart>();
 
-                var uploadId = client.createMultipartUpload(CreateMultipartUploadRequest.builder()
-                        .bucket(bucketName)
-                        .key(key)
-                        .build()).uploadId();
-
-                while (true) {
-                    var bytesChunk = input.readNBytes(chunkSize);
-
-                    if (bytesChunk.length < 1) {
-                        break;
+                try {
+                    var uploadId = client.createMultipartUpload(CreateMultipartUploadRequest.builder()
+                            .bucket(bucketName)
+                            .key(key)
+                            .build()).uploadId();
+
+                    while (true) {
+                        var bytesChunk = input.readNBytes(chunkSize);
+
+                        if (bytesChunk.length < 1) {
+                            break;
+                        }
+
+                        completedParts.add(CompletedPart.builder().partNumber(partNumber)
+                                .eTag(client.uploadPart(UploadPartRequest.builder()
+                                        .bucket(bucketName)
+                                        .key(key)
+                                        .uploadId(uploadId)
+                                        .partNumber(partNumber)
+                                        .build(), RequestBody.fromByteBuffer(ByteBuffer.wrap(bytesChunk))).eTag()).build());
+                        partNumber++;
                     }
 
-                    completedParts.add(CompletedPart.builder().partNumber(partNumber)
-                            .eTag(client.uploadPart(UploadPartRequest.builder()
-                                    .bucket(bucketName)
-                                    .key(key)
-                                    .uploadId(uploadId)
-                                    .partNumber(partNumber)
-                                    .build(), RequestBody.fromByteBuffer(ByteBuffer.wrap(bytesChunk))).eTag()).build());
-                    partNumber++;
+                    client.completeMultipartUpload(CompleteMultipartUploadRequest.builder()
+                            .bucket(bucketName)
+                            .key(key)
+                            .uploadId(uploadId)
+                            .multipartUpload(CompletedMultipartUpload.builder()
+                                    .parts(completedParts)
+                                    .build())
+                            .build());
+                } catch (Exception e) {
+                    throw new RuntimeException(String.format("Fail uploading object to the destination bucket: %s", e.getMessage()), e);
```

Comment:
please throw an `EdcException`, and either use static imports for the `format` call, or - even better - leverage `"xyzabc".formatted(...)`

Reply:
Rethrown exceptions removed